ngraph/core/include/ngraph/function.hpp

```diff
@@ -51,6 +55,34 @@ namespace ngraph
                  const ParameterVector& parameters,
                  const std::string& name = "");
 
+        Function(const ResultVector& results,
+                 const SinkVector& sinks,
+                 const ParameterVector& parameters,
+                 const VariableVector& variables,
+                 const std::string& name = "");
+
+        Function(const OutputVector& results,
+                 const SinkVector& sinks,
+                 const ParameterVector& parameters,
+                 const VariableVector& variables,
+                 const std::string& name = "");
+
+        Function(const ResultVector& results,
```

Comment:
IMHO, too much constructors, can we have variables without sinks?

Reply:
I agree, but I do not know how to reduce their number.
>> can we have variables without sinks? 
Theoretically, it can be, in case Assign is connected to Result.

Reply:
As nGraph developer I really doesn't know which constructor should I use. Can we try to minimize the number of constructors (introduce constructors with default arguments or somehow else)?

Reply:
To solve confusion with many constructor variants, maybe we can add introduce a kind of 'FunctionBuilder'.
I.e.
`auto builder = FunctionBuilder().results(resultsVector).sinks(sinksVector).params(paramsVector).vars(varsVector).name(name);
`
then function creation will look like

`Function f = builder.build(); or Function f(builder);
`

Reply:
Looks good, @ilyachur what do you think about Mikhail's proposal?

Reply:
Maybe we can make easier:
```
Function(const NodeVector&)
```
And this vector will contain all "leaves" (Parameters, Results, Assign,...).
Maybe we can add such constructor only for new operations.

Reply:
Variable is not ngraph::Node, so
`Function(const NodeVector&, const VariableVector&)`

And it looks like it's just another constructor in a bunch of others.

Reply:
Yes, I have some notes, we just agreed that we need all these constructors.But from the user's point of view, it is difficult to know which c-tor he needs to use.
 I guess this is not a blocker of these changes, and can be done it in the next PR, I suggest to discuss this separately